crates/types/src/spline_motion_interpolator.rs

```diff
@@ -0,0 +1,107 @@
+use splines::{Interpolation, Key, Spline};
+use thiserror::Error;
+
+use crate::{Joints, MotionFile};
+use std::time::Duration;
+
+pub struct SplineMotionFileInterpolator {
+    interpolator: Spline<f32, Joints>,
+    current_time: Duration,
+    start_time: Duration,
+    end_time: Duration,
+}
+
+#[derive(Error, Debug)]
+pub enum InterpolatorError {
+    #[error("accessed interpolator defined for [{min}s, {max}s] at {current}s")]
+    OutOfBoundsAccess { min: f32, max: f32, current: f32 },
+}
+
+impl From<MotionFile> for SplineMotionFileInterpolator {
+    fn from(motion_file: MotionFile) -> Self {
+        assert!(!motion_file.frames.is_empty());
+
+        let mut current_time = Duration::ZERO;
+        let mut keys = vec![Key::new(
+            current_time,
+            motion_file.initial_positions,
+            Interpolation::Linear,
+        )];
+
+        keys.extend(motion_file.frames.into_iter().map(|frame| {
+            current_time += frame.duration;
+            Key::new(current_time, frame.positions, Interpolation::Linear)
+        }));
+
+        Self::new(keys)
+    }
+}
+
+impl SplineMotionFileInterpolator {
+    pub fn new(keys: Vec<Key<Duration, Joints>>) -> Self {
+        assert!(keys.len() >= 2, "need at least two keys to interpolate");
+        //NOTE: assume keys are sorted
```

Comment:
Nobody will remember this comment in a few months. Could you express this in code or guarantee it somehow?

Reply:
Done